### PR TITLE
[TextFields, Themes] Replace unnecessary imports with forward declarations

### PR DESCRIPTION
**Pull request by sigito**
Partially addresses #1523.

### CODE REVIEW

File: components/TextFields/src/ColorThemer/MDCTextFieldColorThemer.h
Comment (romoore) [inline components/TextFields/src/ColorThemer/MDCTextFieldColorThemer.h]: Please add `#import <Foundation/Foundation.h>` for NSObject.

File: components/TextFields/src/private/MDCTextInputAllCharactersCounter.m
Comment (romoore) [inline components/TextFields/src/private/MDCTextInputAllCharactersCounter.m]: Nit: I believe `MDCTextInput.h` should be ordered before `MDCTextInputCharacterCounter.h` (NULL < 'C').  Also please remove the extra newline.

**Review by romoore — DISMISSED** [review]
This looks good to me, thanks for the PR! 😸   Just a couple of small comments.

File: components/TextFields/src/ColorThemer/MDCTextFieldColorThemer.h
Comment (sigito) [reply components/TextFields/src/ColorThemer/MDCTextFieldColorThemer.h]: Should I add a Foundation import everywhere NSObject is used? If UIKit is already imported, should I still add a Foundation import there? WDYT?

File: components/TextFields/src/private/MDCTextInputAllCharactersCounter.m
Comment (sigito) [reply components/TextFields/src/private/MDCTextInputAllCharactersCounter.m]: I put the `MDCTextInputCharacterCounter.h` first, and separated with an empty line, because it is the main declaration import. This is a common practice, that is also used around your code base. Should I still make changes suggested by you?

File: components/TextFields/src/ColorThemer/MDCTextFieldColorThemer.h
Comment (romoore) [reply components/TextFields/src/ColorThemer/MDCTextFieldColorThemer.h]: I think you should import Foundation/UIKit if none of your imports do so transitively.

Comment (sigito) [reply components/TextFields/src/ColorThemer/MDCTextFieldColorThemer.h]: Added an import.

File: components/TextFields/src/private/MDCTextInputAllCharactersCounter.m
Comment (romoore) [reply components/TextFields/src/private/MDCTextInputAllCharactersCounter.m]: Your formatting was correct. I didn't realize that the header declared this class as well.  Sorry!

**Comment by sigito** [comment]
I am working on the rest of the files. It doesn't seem like many changes, so I'll create a single PR for everything.

**Comment by ajsecord** [comment]
I think this is a great set of changes, but will _definitely_ cause a lot of breakage for the release that contains them, so while we're still working hard on supporting iPhone X, I'd like to hold off merging. Don't let that stop you, though, keep rolling and we'll merge once you're all done and the iPhone X changes have settled. Thanks!

**Comment by ianegordon** [comment]
Hi @sigito !  Now that the iPhone X is out, we can revisit this PR.  Could you rebase?  I'm hopeful it will take care of some of the conflicts.

**Comment by sigito** [comment]
@ianegordon , synced.

**Comment by willlarche** [comment]
Hey @sigito ! I think you need to do a rebase and then force push. If you're nervous, you can --dry-run first. Unfortunately, GitHub still thinks a billion files need to change.

**Comment by sigito** [comment]
Who is nervous? 


Should be fine now.

**Comment by ianegordon** [comment]
LGTM, we just need conflicts & Travis sorted.

**Review by ianegordon — APPROVED** [review]
Please fix Conflict & Travis before Merge.

We should explicitly mention this in our next release's changelog and the associated test script.

**Review by ianegordon — CHANGES_REQUESTED** [review]
Travis is complaining ☹️ 

/Users/travis/build/material-components/material-components-ios/components/TextFields/examples/TextFieldInterfaceBuilderLegacyExample.m:19:9: 'supplemental/TextFieldInterfaceBuilderLegacyExampleSupplemental.h' file not found
#import "supplemental/TextFieldInterfaceBuilderLegacyExampleSupplemental.h"

**Comment by sigito** [comment]
Looking at it. The codebase evolves faster than I sync :)

**Comment by sigito** [comment]
I've rebased and fixed the imports. Let's see now.

P.S., I wish I made more than just a single PR with these changes.

**Comment by willlarche** [comment]
The git history got a little confused (GitHub's fault?) which happens on some PRs

Let's close this PR and make a new one and we'll give it approval.

**Comment by sigito** [comment]
#2429.

**Pull request closed without merging**
